bndtools.builder/src/org/bndtools/builder/classpath/BndContainerInitializer.java

```diff
@@ -61,15 +64,25 @@
 
     public BndContainerInitializer() {
         super();
-        Central.getInstance().addModelListener(this);
     }
 
     @Override
-    public void initialize(IPath containerPath, IJavaProject javaProject) throws CoreException {
-        IProject project = javaProject.getProject();
+    public void initialize(IPath containerPath, final IJavaProject javaProject) throws CoreException {
+        Central.onWorkspaceInit(new Function<Workspace,Void>() {
+            @Override
+            public Void apply(Workspace a) {
+                try {
+                    Central.getInstance().addModelListener(BndContainerInitializer.this);
```

Comment:
But the initialize() call is what is blocking the UI thread in some cases.  So we need to let the BndContainerInitializer go ahead and setup what it can (even though many classpath entries wont yet be realized because they need to be downloaded) and then once they are "available" then it can "reinit".